### PR TITLE
feat: async checkpointing for Megatron policy workers

### PR DESCRIPTION
**Pull request by ananthsub**
# What does this PR do ?

Expose Megatron-Bridge checkpoint knobs via a new MegatronCheckpointConfig (`async_save`, `ckpt_assume_constant_structure`, and the parallel-IO fields `ckpt_fully_parallel_save`/`load_process_group`, `ckpt_fully_parallel_load_exchange_algo`) and wire them through `_create_checkpoint_config` into the bridge CheckpointConfig.

With `async_save=True`, `MegatronPolicyWorker.save_checkpoint` returns after D2H staging; the disk write continues in a background worker. New `finalize_async_save()`/`terminate_async_checkpoint_worker()` block until the write completes. On a cold start (no prior checkpoint), save falls back to `pretrained_path`, which the bridge requires when async_save is on (CheckpointConfig.finalize() asserts save is not None).

Checkpoint finalization is handled by a daemon inside CheckpointManager: 
- `begin_finalization()` renames `tmp_step_N` to `step_N` in a background thread once writers finish (via `wait_fn`) and queues old-checkpoint deletion
- `finalize_pending()` flushes before the next save
- `shutdown()` flushes on exit

 
The GRPO drivers use these instead of inline deferred finalization. This behaves identically (e.g. block immediately) when async save is disabled.

Recipes:
- Enable async checkpointing (async_save + ckpt_assume_constant_structure) in the base megatron recipe configs; inheriting recipes pick it up via the defaults merge.

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

### CODE REVIEW

File: nemo_rl/algorithms/grpo.py [hunk 1]
```diff
@@ -2683,10 +2695,12 @@ def grpo_train(
             current_step += 1
             total_steps += 1
             if should_save_by_timeout:
+                checkpointer.shutdown()
```
Comment (yfw) [inline nemo_rl/algorithms/grpo.py:3295]: [`grpo.py:2698`](https://github.com/NVIDIA-NeMo/RL/blob/33a2e506ffa1275aa847e5b7e8d47fb336a54ea2/nemo_rl/algorithms/grpo.py#L2698)

**Likely regression:** `grpo_train` has no `try/finally` around its loop, and `shutdown()` is only called at the two early returns here (timeout) and L2703 (`max_num_steps`). For an **epoch-bounded** run, `is_last_step` goes true at the epoch boundary ([L2320-2323](https://github.com/NVIDIA-NeMo/RL/blob/33a2e506ffa1275aa847e5b7e8d47fb336a54ea2/nemo_rl/algorithms/grpo.py#L2320)), so the final step saves and starts `begin_finalization()` ([L2538](https://github.com/NVIDIA-NeMo/RL/blob/33a2e506ffa1275aa847e5b7e8d47fb336a54ea2/nemo_rl/algorithms/grpo.py#L2538), a `daemon=True` thread) — but neither return fires, the loop exits, and the daemon is killed mid-write. The final `tmp_step_N` is never renamed to `step_N` → the last checkpoint is lost. Same on any mid-loop exception.

Previously `finalize_checkpoint()` renamed synchronously before return, so this is a behavior change. `async_grpo_train` already guards this with `try/finally` + `shutdown()` ([L3857-3862](https://github.com/NVIDIA-NeMo/RL/blob/33a2e506ffa1275aa847e5b7e8d47fb336a54ea2/nemo_rl/algorithms/grpo.py#L3857)); consider mirroring that here.

File: nemo_rl/models/policy/workers/megatron_policy_worker.py [hunk 2]
```diff
@@ -1694,6 +1704,29 @@ def save_checkpoint(
...
+        Safe to call when async_save is disabled (no-op).
+        Does NOT terminate the persistent worker — it stays alive for the next save.
+        """
+        maybe_finalize_async_save(
+            self.mcore_state,
+            ckpt_cfg=self.mcore_state.cfg.checkpoint,
+            blocking=True,
+        )
+
+    def terminate_async_checkpoint_worker(self):
```
Comment (yfw) [inline nemo_rl/models/policy/workers/megatron_policy_worker.py]: [`megatron_policy_worker.py:1719`](https://github.com/NVIDIA-NeMo/RL/blob/33a2e506ffa1275aa847e5b7e8d47fb336a54ea2/nemo_rl/models/policy/workers/megatron_policy_worker.py#L1719)

`terminate_async_checkpoint_worker` has no callers anywhere in `nemo_rl/` or `examples/` — the persistent async-worker subprocess is never explicitly closed and relies on Ray actor teardown. Also, the docstring says it "blocks until any in-flight write completes, then shuts down," but the body only calls `async_calls_queue.close()` (no separate blocking finalize). Either wire it into a shutdown path (e.g. alongside `policy.shutdown()`) or remove it.

File: nemo_rl/models/megatron/setup.py [hunk 3]
```diff
@@ -787,20 +788,50 @@ def _create_checkpoint_config(
...
+
+    Args:
+        pretrained_path: Path to the pretrained checkpoint.
+        weights_path: Path to save/load training weights.
+        optimizer_path: Path to the optimizer state (None if not resuming optimizer).
+        load_main_params_from_ckpt: Load optimizer main params from the checkpoint.
+        ckpt_cfg: Optional MegatronCheckpointConfig dict from YAML.
+    """
+    cfg = ckpt_cfg or {}
+    async_save = cfg.get("async_save", False)
```
Comment (yfw) [inline nemo_rl/models/megatron/setup.py]: [`setup.py:803`](https://github.com/NVIDIA-NeMo/RL/blob/33a2e506ffa1275aa847e5b7e8d47fb336a54ea2/nemo_rl/models/megatron/setup.py#L803)

[config-conventions](https://github.com/NVIDIA-NeMo/RL/blob/main/.agents/contributor-skills/config-conventions/SKILL.md): `async_save = cfg.get("async_save", False)` (and L824 `ckpt_assume_constant_structure=cfg.get(..., False)`) is the forbidden call-site-default pattern — the skill's own bad example is `cfg.get("disable_ppo_ratio", False)`. `MegatronConfig` / `PolicyConfig` / `MegatronCheckpointConfig` are all v1 `TypedDict` consumed as a plain dict (`config["megatron_cfg"][...]`), and sibling fields like `sequence_parallel` ([L562](https://github.com/NVIDIA-NeMo/RL/blob/33a2e506ffa1275aa847e5b7e8d47fb336a54ea2/nemo_rl/models/megatron/setup.py#L562)) are already read with no fallback — so the exemplar YAML is the only thing that supplies a runtime default here. (A `BaseModel` field default wouldn't take effect, since nothing instantiates the model on this dict-shaped subtree.)

**Recommended (fully convention-compliant):**
1. Add a documented `checkpoint:` block (`async_save: false`, `ckpt_assume_constant_structure: false`) under `megatron_cfg:` to each **root exemplar** that has a `megatron_cfg`. Since `_create_checkpoint_config` runs for every Megatron algo, that's the 6 roots: `grpo_math_1B.yaml` (covers all grpo/vlm/gdpo/audio + their `_megatron`/`_fp8`/8B/70B/qwen30ba3b variants via `defaults:`), `dpo.yaml`, `rm.yaml`, `sft.yaml`, `ppo_math_1B_megatron.yaml`, `distillation_math.yaml`. The inheriting variants don't need their own copy.
2. Drop the `cfg.get(..., False)` fallbacks at L803/L824 and read the values directly.
3. If any of those roots are covered by `tests/unit/test_config_v2.py`, update the matching `tests/unit/reference_configs/*.yaml` so the v1→v2 migration check stays green.

Note: the megatron recipes in this PR all inherit from `grpo_math_1B.yaml`, so that root is the most important; the other 5 keep the non-GRPO Megatron configs consistent once the call-site fallback is removed.

Comment (yfw) [reply nemo_rl/models/megatron/setup.py]: @ananthsub does it make sense to default this to True?

**Comment by terrykong** [comment]
@ashors1 to review

File: nemo_rl/models/megatron/setup.py [hunk 4]
```diff
@@ -909,20 +910,59 @@ def _create_checkpoint_config(
...
+
+    kwargs: dict[str, Any] = dict(
         save_interval=100,
-        save=weights_path,
+        save=save_path,
         load=weights_path,
         load_optim=optimizer_path is not None,
         pretrained_checkpoint=pretrained_path,
-        async_save=False,
+        async_save=async_save,
```
Comment (ashors1) [inline nemo_rl/models/megatron/setup.py]: [`setup.py:949`](https://github.com/NVIDIA-NeMo/RL/blob/d68347b178f36407e8c191d7b9c27ac3ec6584a1/nemo_rl/models/megatron/setup.py#L949)

**Async save is exposed for every Megatron algorithm, but only GRPO finalizes it safely.** `_create_checkpoint_config` sets `async_save` generically, and with `async_save=True` `MegatronPolicyWorker.save_checkpoint` returns after D2H staging while the disk write continues in the background (the async path does not block at the end — [worker L1657/L1696](https://github.com/NVIDIA-NeMo/RL/blob/d68347b178f36407e8c191d7b9c27ac3ec6584a1/nemo_rl/models/policy/workers/megatron_policy_worker.py#L1696)).

Only `grpo_train`/`async_grpo_train` wait for that write before renaming, via [`begin_finalization(wait_fn=policy.finalize_async_save)`](https://github.com/NVIDIA-NeMo/RL/blob/d68347b178f36407e8c191d7b9c27ac3ec6584a1/nemo_rl/algorithms/grpo.py#L3128). Every other Megatron driver renames immediately with the synchronous `finalize_checkpoint()` and never calls `finalize_async_save`:
- [`sft.py:594`](https://github.com/NVIDIA-NeMo/RL/blob/d68347b178f36407e8c191d7b9c27ac3ec6584a1/nemo_rl/algorithms/sft.py#L594)
- [`dpo.py:736`](https://github.com/NVIDIA-NeMo/RL/blob/d68347b178f36407e8c191d7b9c27ac3ec6584a1/nemo_rl/algorithms/dpo.py#L736)
- [`rm.py:671`](https://github.com/NVIDIA-NeMo/RL/blob/d68347b178f36407e8c191d7b9c27ac3ec6584a1/nemo_rl/algorithms/rm.py#L671)
- [`distillation.py:1029`](https://github.com/NVIDIA-NeMo/RL/blob/d68347b178f36407e8c191d7b9c27ac3ec6584a1/nemo_rl/algorithms/distillation.py#L1029)
- [`ppo.py:1604`](https://github.com/NVIDIA-NeMo/RL/blob/d68347b178f36407e8c191d7b9c27ac3ec6584a1/nemo_rl/algorithms/ppo.py#L1604)
- [`grpo_sync.py:1161`](https://github.com/NVIDIA-NeMo/RL/blob/d68347b178f36407e8c191d7b9c27ac3ec6584a1/nemo_rl/algorithms/grpo_sync.py#L1161)
- [`xtoken_off_policy_distillation.py:782`](https://github.com/NVIDIA-NeMo/RL/blob/d68347b178f36407e8c191d7b9c27ac3ec6584a1/nemo_rl/algorithms/xtoken_off_policy_distillation.py#L782)

So `async_save: true` on any of these renames `tmp_step_N` -> `step_N` while the write is still in flight -> a silently corrupt/incomplete checkpoint. The shipped exemplars all default these to `false`, so it only bites a user who opts in, but nothing guards it. Per error-handling (*fail loud, not silent*), consider a setup-time `raise` rejecting `async_save: true` for algorithms whose driver uses synchronous `finalize_checkpoint()`, or route those drivers through the async-aware `begin_finalization` path.

File: nemo_rl/models/megatron/setup.py [hunk 5]
```diff
@@ -909,20 +910,59 @@ def _create_checkpoint_config(
...
+            ``async_save: true`` explicitly. ``async_save`` defaults to ``False`` when
+            absent (sync save, pre-existing behavior) because nemo-rl uses it for
+            save-path logic as well as passing through to Megatron Bridge.
+            ``ckpt_assume_constant_structure`` is only forwarded when set in YAML;
+            otherwise Megatron Bridge's own default applies.
+    """
+    cfg = ckpt_cfg or {}
+    # async_save is always resolved: nemo-rl branches on it (save-path fallback) and
+    # passes it to CheckpointConfig. Default False preserves pre-existing sync behavior.
+    async_save = cfg.get("async_save", False)
```
Comment (ashors1) [inline nemo_rl/models/megatron/setup.py]: [`setup.py:933`](https://github.com/NVIDIA-NeMo/RL/blob/d68347b178f36407e8c191d7b9c27ac3ec6584a1/nemo_rl/models/megatron/setup.py#L933)

`async_save = cfg.get("async_save", False)` is the call-site-default pattern that [config-conventions](https://github.com/NVIDIA-NeMo/RL/blob/main/.agents/contributor-skills/config-conventions/SKILL.md) forbids (its own bad example is `cfg.get("disable_ppo_ratio", False)`). The `checkpoint:` block was correctly added to all 6 exemplars + reference_configs, so the exemplar YAML can own the default — you can drop the `, False` and read it as the sibling `ckpt_assume_constant_structure` is already read ([L955-961](https://github.com/NVIDIA-NeMo/RL/blob/d68347b178f36407e8c191d7b9c27ac3ec6584a1/nemo_rl/models/megatron/setup.py#L955), presence-checked, no fallback).

Note: converting `MegatronCheckpointConfig` to a `BaseModel` would *not* let the default live on the field here, since the parent `MegatronConfig` is a TypedDict consumed as a plain dict (`config["megatron_cfg"].get("checkpoint")`) and is never instantiated — so the YAML remains the source of truth. This is @yfw's still-open thread; the "default to True?" question also looks unanswered (exemplars use `true`, the code fallback is `False`).

File: nemo_rl/models/megatron/setup.py [hunk 6]
```diff
@@ -1164,6 +1204,18 @@ def setup_model_and_optimizer(
...
+    # Must be called before initialize_megatron (before CUDA init) so the
+    # persistent async-checkpoint worker subprocess is spawned in a clean process.
+    # Bridge hardcodes mp_mode='spawn', the only safe option inside Ray actors
+    # (fork with Ray is an anti-pattern). Guard for older megatron-core that
+    # lacks this hook.
+    _init_async_ckpt_worker = getattr(state, "initialize_async_checkpoint_worker", None)
+    if _init_async_ckpt_worker is not None:
+        try:
+            _init_async_ckpt_worker()
+        except AttributeError:
```
Comment (ashors1) [inline nemo_rl/models/megatron/setup.py]: [`setup.py:1216`](https://github.com/NVIDIA-NeMo/RL/blob/d68347b178f36407e8c191d7b9c27ac3ec6584a1/nemo_rl/models/megatron/setup.py#L1216)

The `getattr(state, "initialize_async_checkpoint_worker", None)` guard above already handles older megatron-core that lacks the hook, so this inner `except AttributeError: pass` can only swallow an `AttributeError` raised *inside* the method — e.g. from [`get_async_strategy` / `async_modules[...]` / `warmup_persistent_caller`](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/554c7b9324225aa863eee52e8b8fdde7abced2b1/src/megatron/bridge/training/state.py#L416-L430). If that happens the persistent async worker silently fails to spawn before CUDA init (the exact spawn/fork hazard the comment warns about), with no error. Recommend dropping the inner `try/except` and relying on the `getattr` guard. (error-handling: fail loud.)

File: nemo_rl/utils/checkpoint.py [hunk 7]
```diff
@@ -229,37 +246,103 @@ def init_tmp_checkpoint(
...
+        self.finalize_pending()
+        self._pending_checkpoint_path = Path(checkpoint_path)
+        self._finalize_error = None
+
+        def _finalize():
+            try:
+                if wait_fn is not None:
+                    wait_fn()
+                self._rename_checkpoint(checkpoint_path)
+                self._delete_executor.submit(self.remove_old_checkpoints)
```
Comment (ashors1) [inline nemo_rl/utils/checkpoint.py]: [`checkpoint.py:309`](https://github.com/NVIDIA-NeMo/RL/blob/d68347b178f36407e8c191d7b9c27ac3ec6584a1/nemo_rl/utils/checkpoint.py#L309)

The `Future` from `_delete_executor.submit(self.remove_old_checkpoints)` is discarded. [`shutdown()`](https://github.com/NVIDIA-NeMo/RL/blob/d68347b178f36407e8c191d7b9c27ac3ec6584a1/nemo_rl/utils/checkpoint.py#L332) does `_delete_executor.shutdown(wait=True)`, which waits but never re-raises task exceptions. If `remove_old_checkpoints` fails (malformed `training_info.json`, or an `rmtree` racing a reader) top-k pruning silently stops and disk fills with no signal — unlike the synchronous [`finalize_checkpoint`](https://github.com/NVIDIA-NeMo/RL/blob/d68347b178f36407e8c191d7b9c27ac3ec6584a1/nemo_rl/utils/checkpoint.py#L268) where it would propagate. Consider retaining the future and checking/logging it in `finalize_pending()`/`shutdown()`.

File: nemo_rl/algorithms/grpo.py [hunk 8]
```diff
@@ -2336,917 +2336,979 @@ def grpo_train(
...
+
+            current_epoch += 1
+            current_step = 0  # Reset step counter for new epoch
+    finally:
+        # Flush the last checkpoint's background finalization before exiting.
+        # A normal loop exit (epoch limit reached) or a mid-loop exception
+        # bypasses the inline shutdown() calls at the early returns; without
+        # this flush the daemon thread is killed before it renames the final
+        # tmp_step_N directory.
+        checkpointer.shutdown()
```
Comment (ashors1) [inline nemo_rl/algorithms/grpo.py]: [`grpo.py:3311`](https://github.com/NVIDIA-NeMo/RL/blob/d68347b178f36407e8c191d7b9c27ac3ec6584a1/nemo_rl/algorithms/grpo.py#L3311)

This `finally: checkpointer.shutdown()` is bare. If the loop raised a real training error *and* the final finalization also fails, `shutdown()` re-raises `RuntimeError("Background checkpoint finalization failed")` and masks the original exception. `async_grpo_train` already guards this ([L4517-4520](https://github.com/NVIDIA-NeMo/RL/blob/d68347b178f36407e8c191d7b9c27ac3ec6584a1/nemo_rl/algorithms/grpo.py#L4517) wraps in try/except + print).

**Optional, cleaner alternative:** make `CheckpointManager` a context manager (`__enter__`/`__exit__`, where `__exit__` runs `shutdown()` but never lets a finalization error mask an in-flight training exception). Then both drivers become `with checkpointer:` around the loop, the two redundant inline [`shutdown()`](https://github.com/NVIDIA-NeMo/RL/blob/d68347b178f36407e8c191d7b9c27ac3ec6584a1/nemo_rl/algorithms/grpo.py#L3288) calls at the early returns can go away, and the `grpo_train` vs `async_grpo_train` asymmetry disappears — the fix lives in one place.

File: nemo_rl/utils/checkpoint.py [hunk 9]
```diff
@@ -229,37 +246,103 @@ def init_tmp_checkpoint(
...
+        step = checkpoint_path.name.split("_")[2]
+        to_checkpoint_path = checkpoint_path.parent / f"step_{step}"
         if to_checkpoint_path.exists():
-            # if step_{step} exists, rename it to old_step_{step}, move tmp_step_{step} to step_{step}, then delete
-            # we do this trickery to have a 'pseudo-atomic' checkpoint save
-            old_checkpoint_path = (
-                checkpoint_path.parent
-                / f"old_step_{checkpoint_path.name.split('_')[2]}"
-            )
+            old_checkpoint_path = checkpoint_path.parent / f"old_step_{step}"
```
Comment (ashors1) [inline nemo_rl/utils/checkpoint.py:260]: [`checkpoint.py:259`](https://github.com/NVIDIA-NeMo/RL/blob/d68347b178f36407e8c191d7b9c27ac3ec6584a1/nemo_rl/utils/checkpoint.py#L259)

The `step_N`-already-exists swap branch (`old_step_N` pseudo-atomic swap — the resume edge case the code explicitly calls out) is untested; every current test finalizes a given step exactly once. Verified-passing test to close the gap (add to `tests/unit/utils/test_checkpoint.py`):

```python
def test_rename_checkpoint_swaps_when_target_exists(checkpoint_manager, checkpoint_dir):
    """_rename_checkpoint swaps in-place when step_N already exists (resume edge case)."""
    tmp1 = checkpoint_manager.init_tmp_checkpoint(1, {"loss": 0.1})
    checkpoint_manager.finalize_checkpoint(tmp1)
    assert (checkpoint_dir / "step_1").exists()

    tmp2 = checkpoint_manager.init_tmp_checkpoint(1, {"loss": 0.999})
    checkpoint_manager.finalize_checkpoint(tmp2)

    assert (checkpoint_dir / "step_1").exists()
    assert not (checkpoint_dir / "old_step_1").exists()
    assert not Path(tmp2).exists()
    metadata = checkpoint_manager.load_training_info(checkpoint_dir / "step_1")
    assert metadata["loss"] == 0.999
```

File: nemo_rl/utils/checkpoint.py [hunk 10]
```diff
@@ -229,37 +246,103 @@ def init_tmp_checkpoint(
...
+        """Block until rename + all queued deletions complete. Call at training exit.
+
+        Safe to call multiple times.
+        """
+        self.finalize_pending()
+        self._delete_executor.shutdown(wait=True)
+        self._delete_executor = ThreadPoolExecutor(max_workers=1)
+
+    @property
+    def has_pending_finalization(self) -> bool:
```
Comment (ashors1) [inline nemo_rl/utils/checkpoint.py:388]: [`checkpoint.py:342`](https://github.com/NVIDIA-NeMo/RL/blob/d68347b178f36407e8c191d7b9c27ac3ec6584a1/nemo_rl/utils/checkpoint.py#L342)

`has_pending_finalization` has no production caller — it's referenced only by its own tests in `tests/unit/utils/test_checkpoint.py`. The driver loops drive ordering explicitly via `finalize_pending()`/`shutdown()`. Low-severity: either remove it (and its tests), or if it's intended as a public status hook for future callers (logging/metrics), note that in the docstring.

File: nemo_rl/models/policy/__init__.py [hunk 11]
```diff
@@ -237,6 +237,21 @@ class MegatronConfigDisabled(TypedDict):
     enabled: Literal[False]
 
 
+class MegatronCheckpointConfig(TypedDict, total=False):
```
Comment (ashors1) [inline nemo_rl/models/policy/__init__.py:240]: [`__init__.py:240`](https://github.com/NVIDIA-NeMo/RL/blob/d68347b178f36407e8c191d7b9c27ac3ec6584a1/nemo_rl/models/policy/__init__.py#L240)

The new user-facing `megatron_cfg.checkpoint` knobs (`async_save`, `ckpt_assume_constant_structure`, the parallel-IO fields) ship without `docs/` coverage (the PR checklist doc box is unchecked). Optional: a short checkpointing section documenting the knobs and the finalize-on-exit semantics would help users adopt it.

**Review by ashors1 — COMMENTED** [review]
Reviewed with a team of agents (RL/config expert, bug-finder, test-reviewer, comment-reviewer) plus an adversarial verification pass. Nice feature — the async finalization design is sound, the new unit tests are thorough (37 pass locally), the cold-start `save=pretrained_path` fallback is correct, and both of @yfw's prior blocking threads are resolved.

One higher-severity item worth a look before merge: **async save is wired generically for all Megatron algorithms, but only GRPO's driver waits for the background write before renaming** — see the inline comment on `setup.py`. The remaining inline comments are low-severity suggestions (a config-conventions call-site default, two fail-silent smells, a test-coverage gap, an unused property, plus an optional docs ask).

Verified independently: the finalize-thread race conditions are safe (single daemon serialized via `join()`), and the cold-start fallback never overwrites the pretrained checkpoint (`save_checkpoint` sets the real path at worker L1667 before every write).

_Generated by Claude Code_

File: nemo_rl/models/megatron/setup.py (hunk 5 above)
Comment (ananthsub) [reply nemo_rl/models/megatron/setup.py]: i'd like to keep the optional fallback as False to avoid this being a breaking change for anyone using a config that doesn't inherit from the exemplars

Comment (terrykong) [reply nemo_rl/models/megatron/setup.py]: i think the precedent so far has been to add it to the exemplar and avoid defaults (i understand it can be a little jarring)

there's another work stream led by @yuki-97 to move us to pydantic basemodels which should ultimately clean this all up even for ppl not using our exemplar configs. i think for now it's better to just stick to the convention

**Review by ashors1 — APPROVED** [review]
